src/trajectory/BSpline.cpp

```diff
@@ -0,0 +1,499 @@
+}
+
+//==============================================================================
+double BSpline::getStartTime() const
+{
+  return mStartTime;
+}
+
+//==============================================================================
+double BSpline::getEndTime() const
+{
+  return mEndTime;
+}
+
+//==============================================================================
+double BSpline::getDuration() const
+{
+  return mEndTime - mStartTime;
+}
+
+//==============================================================================
+void BSpline::evaluate(double t, statespace::StateSpace::State* state) const
+{
+  Eigen::VectorXd values(mStateSpace->getDimension());
+
+  for (auto i = 0u; i < mStateSpace->getDimension(); ++i)
+    values[i] = mSplines[i](t)[0];
+
+  mStateSpace->expMap(values, state);
+}
+
+//==============================================================================
+void BSpline::evaluateDerivative(
+    double /*t*/, int derivative, Eigen::VectorXd& /*tangentVector*/) const
+{
+  if (derivative < 1)
+    throw std::logic_error("Derivative must be positive.");
+
+  // TODO(JS): Not implemented
+}
+
+//==============================================================================
+double BSpline::computeArcLength(
+    const distance::DistanceMetric& distanceMetric, double resolution) const
+{
+  const common::StepSequence steps(
+      resolution, true, true, getStartTime(), getEndTime());
+
+  if (steps.getLength() == 0u)
+    return 0.0;
+
+  auto state0 = mStateSpace->createState();
+  auto state1 = mStateSpace->createState();
+
+  double arcLength = 0.0;
+
+  auto it = steps.begin();
+  evaluate(*it, state0);
+  for (; it != steps.end(); ++it)
+  {
+    evaluate(*it, state1);
+    arcLength += distanceMetric.distance(state0, state1);
+
+    mStateSpace->copyState(state1, state0);
+  }
+
+  return arcLength;
+}
+
+//==============================================================================
+std::size_t BSpline::computeNumKnots(
+    std::size_t degree, std::size_t numControlPoints)
+{
+  return degree + numControlPoints + 1u;
+}
+
+//==============================================================================
+BSpline::KnotVectorType BSpline::computeUniformKnots(
+    std::size_t degree,
+    std::size_t numControlPoints,
+    double startTime,
+    double endTime)
+{
+  if (numControlPoints < degree + 1u)
+  {
+    throw std::invalid_argument(
+        "Number of control points should be equal to or greater than degree + "
+        "1");
+  }
+
+  const auto numKnots = computeNumKnots(degree, numControlPoints);
+
+  KnotVectorType knots(numKnots);
+  knots.head(static_cast<ControlPointVectorType::Index>(degree))
+      .setConstant(startTime);
+  knots.tail(static_cast<ControlPointVectorType::Index>(degree))
+      .setConstant(endTime);
+
+  const common::StepSequence internalKnots(
+      startTime, endTime, numControlPoints + 1u - degree, true);
+  ControlPointVectorType::Index index = 1;
```

Comment:
why `index`==0 is not updated?

Reply:
This line should be fixed to `index = degree`. We use multiple knots of the number of degree at the start and end knots so that the start and end of the trajectory are placed at the first and last control points, respectively. Let me document this to be clear.